Added staff role on ccx to all the staff members of master course.

amir-qayyum-khan opened the pull request:
### Background
From issue https://github.com/mitocw/edx-platform/issues/126

### What is done in this PR
**Studio Updates:** None.

**LMS Updates:** 
- Now when a coach creates a new ccx, I am assigning staff role on this ccx course to all staff of master course.
- Now staff of master course has staff rights on master course as well as ccx courses.
- This is for course staff to see what ccx coaches are doing with course.
- Move code that is not rendering view to utility.

@pdpinch @giocalitri @pwilkins 

- Staff on Master course
![screen shot 2015-11-30 at 5 12 54 pm](https://cloud.githubusercontent.com/assets/10431250/11471231/ba5e9cc8-9785-11e5-95c5-0cdcea97f780.png)

- Staff on CCX
![screen shot 2015-11-30 at 5 13 04 pm](https://cloud.githubusercontent.com/assets/10431250/11471234/c2d0a9dc-9785-11e5-865f-a1b7d79a34db.png)

amir-qayyum-khan commented on common/djangoapps/student/roles.py:
```diff
@@ -47,9 +47,11 @@ def has_role(self, role, course_id, org):
         return any(
             access_role.role == role and
-            access_role.course_id == course_id and
+            unicode(access_role.course_id) == unicode(course_id) and
```
Because code was not able to compare ccxLocator objects though it was able to compare courseLocator.
In sum when this line was comparing ccx keys it always give false even both key are match. Thats why I am deserialising keys and comparing them as strings.

second option is block this PR and create small PR against 
https://github.com/edx/ccx-keys/blob/master/ccx_keys/locator.py

and override eq method. Because we dont have this method thats why when i compare 2 ccx locator objects it always returns false

After that resume this PR
http://stackoverflow.com/questions/1227121/compare-object-instances-for-equality-by-their-attributes-in-python

pdpinch replied:
I don't see a corresponding __eq__ method in https://github.com/edx/opaque-keys/blob/master/opaque_keys/edx/locator.py -- how is this dealt with elsewhere in edX?

pdpinch commented on lms/djangoapps/ccx/views.py:
```diff
@@ -560,6 +564,28 @@ def ccx_course(ccx_locator):
+def add_master_course_staff_to_ccx(course, ccx_key, display_name):
+    """
+    Assign staff of master course, staff role on ccx.
+    :param course: Master course
```
is this a master course key?

amir-qayyum-khan replied:
No it is master course object

pdpinch commented on lms/djangoapps/instructor/access.py:
```diff
@@ -88,7 +88,10 @@ def _change_access(course, user, level, action):
                 email_params=email_params,
             )
-        role.add_users(user)
+        print "role.has_user(user): ", role.has_user(user)
```
debugging code?

pdpinch commented on lms/djangoapps/instructor/access.py:
```diff
@@ -88,7 +88,10 @@ def _change_access(course, user, level, action):
+        print "role.has_user(user): ", role.has_user(user)
+        if not role.has_user(user):
+            # If user does not have this role, only then assign him.
+            role.add_users(user)
```
why is this necessary?

amir-qayyum-khan commented:
@pdpinch fixed issues

amir-qayyum-khan commented:
@pdpinch should i create this on upstream?

pdpinch commented:
Ok. Can you add "wip" to the title, until someone from our team can give it a thumbs up? 

----------------
I am mobile.

> On Dec 1, 2015, at 8:04 AM, Amir Qayyum Khan <notifications@github.com> wrote:
> 
> @pdpinch should i create this on upstream?
> 
> —
> Reply to this email directly or view it on GitHub.
>

giocalitri commented on lms/djangoapps/ccx/views.py:
```diff
@@ -572,6 +576,28 @@ def ccx_course(ccx_locator):
+    Assign staff of master course, staff role on ccx.
+    :param course: Master course instance
+    :param ccx_key: CCX course key
+    :param display_name: ccx display name for email
```
I'm not sure this is the right or current way they want the docstrings.

giocalitri commented on lms/djangoapps/ccx/views.py:
```diff
@@ -572,6 +576,28 @@ def ccx_course(ccx_locator):
 
+def add_master_course_staff_to_ccx(course, ccx_key, display_name):
+    """
+    Assign staff of master course, staff role on ccx.
```
to whom?

giocalitri commented on lms/djangoapps/ccx/views.py:
```diff
@@ -572,6 +576,28 @@ def ccx_course(ccx_locator):
+        email_params = get_email_params(course, auto_enroll=True, course_key=ccx_key, display_name=display_name)
+        for staff in list_staff:
+            allow_access(course, staff, 'staff')
+            # Enroll the coach in the course
```
you are enrolling not only the coach, but any other staff member, if I understood well

giocalitri commented on lms/djangoapps/ccx/views.py:
```diff
@@ -572,6 +576,28 @@ def ccx_course(ccx_locator):
+                email_params=email_params,
+            )
+
+
```
I think this function looks fine.

giocalitri replied:
I think this function looks fine, but what I meant in my comment in the story was to put this function in an external module ( `api.py` ?) so I can import it easily.

amir-qayyum-khan commented:
Done @giocalitri

giocalitri commented on lms/djangoapps/ccx/api.py:
```diff
@@ -0,0 +1,33 @@
+    Added staff role on ccx to all the staff members of master course. 
+    :param course: Master course instance
+    :param ccx_key: CCX course key
+    :param display_name: ccx display name for email
```
Sorry to insist on this, but I think the way they want docstrings now is something like this:

```
Args:
    course (<the object type here>): Master course instance
    ccx_key (<the object type here>): CCX course key
    ...

Returns:
    None  # Not sure this is actually necessary
```

giocalitri replied:
Actually I have seen both, so I am not sure

giocalitri commented on lms/djangoapps/ccx/api.py:
```diff
@@ -0,0 +1,33 @@
+    :param display_name: ccx display name for email
+    """
+    # To avoid circular import
+    from views import ccx_course
```
you should move this outside the `views` file as well

giocalitri replied:
And if there are tests related, we should also move the tests to another file

giocalitri commented:
I know you put the test in the `test_views.py`, but I feel like we need another one in `test_apis.py`

amir-qayyum-khan commented:
@giocalitri Done with moving code and creating test file.
Removed ``api.py`` because the function i made is not api but utility function.
So i moved it in utility.
Updated comment as well.

CC @pdpinch

giocalitri commented on lms/djangoapps/ccx/tests/test_utils.py:
```diff
@@ -3,17 +3,28 @@
 
 from ccx_keys.locator import CCXLocator
+from lms.djangoapps.ccx.overrides import override_field_for_ccx
+from ..utils import add_master_course_staff_to_ccx, ccx_course
```
can please you use the full path for this? Even if it is correct I would avoid using this kind of imports unless the module is in the same directory of the file that imports it

giocalitri commented:
I think you did a great job about moving all the functions in the new module.

amir-qayyum-khan commented:
Made upstream PR https://github.com/edx/edx-platform/pull/10877
